controllers/healthcheck.go

```diff
@@ -0,0 +1,68 @@
+package controllers
+
+//go:generate mockgen -source=healthcheck.go -destination=../mocks/healthcheck.go -package=mocks
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/prest/prest/adapters/postgres"
+)
+
+type HealthCheck struct {
+	Status string `json:"status"`
+}
+
+type DbConnection interface {
+	GetConnection() (*sqlx.DB, error)
+	RunTestQuery() error
+}
+
+type DBConn struct{}
+
+func (d DBConn) GetConnection() (db *sqlx.DB, err error) {
+	db, err = postgres.Get()
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (d DBConn) RunTestQuery() (err error) {
+	db, err := d.GetConnection()
+
+	if err != nil {
+		return err
+	}
+
+	_, err = db.Exec(";")
+
+	return err
+}
+
+func WrappedHealthCheck(dbc DbConnection) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		data := HealthCheck{
+			Status: "ok",
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		_, err := dbc.GetConnection()
```

Comment:
Also if `RunTestQuery()` also uses this method and return its possible error you could do this closure there and remove this call

Reply:
@arxdsilva I did this with the intent of showing two possible errors: Code execution error (a querying error) and a connection error. If you think this is a deal-breaker, I can remove it.

Reply:
I am afraid @vmesel didn't get @arxdsilva's point — or maybe I got it wrong. Anyway… what I have in mind is that the closure and the close call would look like this:

```go
conn, err := func() {
	db, err = postgres.Get()
	if err != nil {
		return nil, err
	}
    return db, nil
}()
if err != nil {
    // …
}
defer conn.Close()
```

Which could be simplified to:

```go
conn, err := func() {
	return postgres.Get()
}()
if err != nil {
    // …
}
defer conn.Close()
```

Which is the same as:

```go
conn, err := postgres.Get()
if err != nil {
    // …
}
defer conn.Close()
```

There are two really important things going on here. I agree with one and disagree with the other.

💯 on the `defer conn.Close()` — we need that.

But working with the closure that way would make it difficult to test different scenarios using mock or dependency injection. So I'd be more careful and investigate further the testing options there.

Finally, Vini, this structure doesn't affect your objective of showing two possible errors.

Reply:
I think I was in need to take a cup of caffeine. Thank you for the explanation, @cuducos. It doesn't mess up, I was thinking of single error types for the same function but this does not need to be true. Now I got it

Reply:
@vmesel @cuducos I believe that we can further simplify those function to something like this:

```go
func (d DBConn) RunTestQuery() (err error) {
	db, err := d.GetConnection()
	if err != nil {
		return err
	}
	defer db.Close() // please check if exists

	_, err = db.Exec(";")
	return err
}

func WrappedHealthCheck(dbc DbConnection) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		// this is abstracted and much more testable/mockable 
		// since it doesnt know any db logic
		err := dbc.RunTestQuery() 
		if err != nil {
			http.Error(w, "unable to run queries on the database", http.StatusServiceUnavailable)
			return
		}

		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(HealthCheck{
			Status: "ok",
		})
	}
}

```

OBS: added comments just to explain more, but this small change should fix the issue, maybe even we could on another step think about making this getConnection private, but I dont mind to keep it that way.